docs/architecture/adr-002-ibc-relayer.md

```diff
@@ -0,0 +1,644 @@
+
+#### IBC Messages
+The relayer creates transactions that include IBC messages to manage clients, connections and channels, and send application packets to destination chains. These messages must be defined in the IBC Rust implementation [IBC-Modules-Rust].
+
+#### IBC Logging System
+IBC packet data & timeouts are not stored directly in the chain state (as this storage is presumed to be expensive) but are instead committed to with a succinct cryptographic commitment (only the commitment is stored).
+As a consequence, IBC requires that a **host state machine MUST provide an event logging system** that logs data in the course of transaction execution. **Logs must be queryable** by relayers to read IBC packet data & timeouts. 
+
+The logging system must provide the following functions:
+ - [IBC-Modules-Go] emitLogEntry for emitting log entries called by the state machine during transaction execution:
+   - type emitLogEntry = (topic: string , data: []byte) => void
+   - example: emitLogEntry("sendPacket", {sequence: packet.sequence , data: packet.data, timeout: packet.timeout})
+ - [IBC-Modules-Go] queryByTopic for querying past logs matching a given topic:
+   - type queryByTopic = (height: uint64 , topic: string) => Array < []byte >
+
+#### Keyring 
+The relay process must have access to its accounts with tokens on all destination chains, with sufficient balance to pay for transaction fees. Account key information must be stored and managed securely. A keyring implementation is required for CRUD key operations. 
+[Keyring-Rust] Investigation in existing Rust implementations is needed. (ex: [hwchen-keyring](https://github.com/hwchen/keyring-rs))
+
+### Chain Transactions and Signing
+The relayer must create chain specific signed transactions. 
+[Cosmos-Tx-Rust] For the first release Cosmos-SDK transaction signing is required. One possible implementation is [iqlusion's sdtx crate](https://github.com/iqlusioninc/crates/tree/develop/stdtx)
+
+#### Implementation of IBC "routing module"
+The default IBC handler uses a receiver call pattern, where modules must individually call the IBC handler in order to bind to
+ports, start handshakes, accept handshakes, send and receive packets, etc. While this provides flexibility for modules, it imposes extra work on the part of the relayer processes that now needs to track the state of multiple modules. The IBC specification describes an IBC “routing module” to route packets, and simplify the task of relayers. This routing module accepts external datagrams and calls into the IBC handler to deal with handshakes and packet relay. The routing module keeps a lookup table of modules, which it can use to look up and call a module when a packet is received, so that external relayers need only ever relay packets to the routing module.
```

Comment:
"so that external relayers need only ever relay packets to the routing module". I don't understand this sentence: I guess that relayer has option of sending transaction to a chain, not to a particular module. Once transaction with IBC packet is received, it goes to routing module that is then forwarding it to the right module. Is this true? And who is responsible for determining if transaction contains IBC packet data?

Reply:
> Is this true? 

Yes
> And who is responsible for determining if transaction contains IBC packet data?

This depends on the chain implementation. In Cosmos-SDK, for application packets, it's going to be the implementation of ICS-026 (this routing module) that keeps track of the different application modules, their channels and ports, and do the correct handling. For handshake messages this is going to be the IBC modules.